### PR TITLE
obsservice: ingest events

### PR DESCRIPTION
**Pull request by andreimatei**
This patch adds a couple of things:
1. an RPC endpoint to CRDB for streaming out events. This RPC service is
   called by the Obs Service.
2. a library in CRDB for exporting events.
3. code in the Obs Service for ingesting the events and writing them
   into a table in the sink cluster.

The first use of the event exporting is for the system.eventlog events.
All events written to that table are now also exported. Once the Obs
Service takes hold in the future, I hope we can remove system.eventlog.

The events are represented using OpenTelemetry protos. Unfortunately,
I've had to copy over the otel protos to our tree because I couldn't
figure out a vendoring solution. Problems encountered for vendoring are:
1. The repo where these protos live
(https://github.com/open-telemetry/opentelemetry-proto) is not
go-get-able. This means hacks are needed for our vendoring tools.
2. Some of the protos in that repo do not build with gogoproto (they
only build with protoc), because they use the new-ish "optional"
option on some fields. The logs protos that we use in this patch do not
have this problem, but others do (so we'll need to figure something out
in the future when dealing with the metrics proto).  FWIW, the
OpenTelemetry Collector ironically has the same problem (it uses
gogoproto), and it solved it through a sed that changes all the optional
fields to one-of's.
3. Even if you solve the first two problems, the next one is that we
already have a dependency on these compiled protos in our tree
(go.opentelemetry.io/proto/otlp). This repo contains generated code,
using protoc. We need it because of our use of the otlp trace exporter.
Bringing in the protos again, and building them with gogo, results in go
files that want to live in the same package/have the same import path.
So changing the import paths is needed.

Between all of these, I've given up - at least for the moment - and
decided to copy over to our tree the few protos that we actually need.
I'm also changing their import paths. You'll notice that there is a
script that codifies the process of transforming the needed protos from
their otel upstream.

Release note: None
Release justification: Non-production.

### CODE REVIEW

**Comment by andreimatei** [comment]
I plan on adding some tests.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, and @dhartunian)

___
*[`pkg/server/tenant.go` line 547 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7WSn0XChmv4MZhs2R5:-N7WSn0XChmv4MZhs2R6:b-fvqv7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/be97459de9b359a12b81f257b5e426cb65934da8/pkg/server/tenant.go#L547)):*
> ```Go
> 		monitorAndMetrics.rootSQLMemoryMonitor, // memMonitor - this is not "SQL" usage, but we don't have another memory pool,
> 	)
> 	// TODO(andrei): figure out what cluster ID and node ID to use and then call
> ```

hey @knz do you have thoughts on how to get a "cluster ID" and a "node ID" for a tenant?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @andreimatei, and @dhartunian)

___
*[`pkg/server/tenant.go` line 547 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7WSn0XChmv4MZhs2R5:-N7WTaaZ6Wf1IfIk4hSV:b-88fjm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/be97459de9b359a12b81f257b5e426cb65934da8/pkg/server/tenant.go#L547)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

hey @knz do you have thoughts on how to get a "cluster ID" and a "node ID" for a tenant?
</blockquote></details>

This point is too early to get them; they are bound later during server initialization.

The best you can do at this point is take a reference to their _container_  (respectively: `rpcContext.LogicalClusterID`, `baseCfg.IDContainer`).

The value can be sampled after `preStart()` completes (which is under `startTenantInternal()`).

**Review by abarganier — COMMENTED** [review]
This is awesome, nice work. A few comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-:-N7bNSM5-LLdTC3Cm4wD:b-gczo1p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @andreimatei, and @dhartunian)

___
*[`-- commits` line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7alIttBeJBYt6_jUG6):*
:fire: 



_Code quote:_
```
I hope we can remove system.eventlog.
```
___
*[`-- commits` line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7an_l645DFlDu2jaXT):*
to be fair, https://github.com/open-telemetry/opentelemetry-proto says:
> The proto files can be consumed as GIT submodules **or copied and built directly in the consumer project**.

We're just following directions :grin: 



_Code quote:_
```
I've given up
```
___
*[`pkg/obs/event_exporter.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7arYT696mhSwzQshAo:-N7arYT696mhSwzQshAp:bhn8mso) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L43)):*
> ```Go
> // requests from the Observability Service to subscribe to the events stream.
> // Once a subscription is established, new events (published through
> // SendEvent()) are sent to the subscriber.
> ```

Cool - I could see us building cockroachdb receiver & exporter components for the otel collector some day to integrate with this . 



_Suggestion:_
```Go
// EventsServer implements the obspb.ObsServer gRPC service. It responds to
// requests from the Observability Service to subscribe to the events stream.
// Once a subscription is established, new events (published through
// SendEvent()) are sent to the subscriber.
```
___
*[`pkg/obs/event_exporter.go` line 310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bMbzHEu20hVk8P3ir:-N7bMbzHEu20hVk8P3is:b-n3o868) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L310)):*
> ```Go
> 	}
> 
> 	buf := s.buf.mu.events[typ]
> ```

nit: should we check/handle `!ok` with this map access?



_Code quote:_
```Go
buf := s.buf.mu.events[typ]
```
___
*[`pkg/obs/event_exporter.go` line 411 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bLUJjDou0ARrVMdxC:-N7bLUJk6Rcj0Lfdl1S9:bpp3xha) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L411)):*
> ```Go
> 		// If we failed to send, we can't use this subscriber anymore.
> 		//
> 		// TODO(andrei): Figure out how to tolerate errors; we should put the events
> ```

in the meantime, maybe we should log something here.



_Code quote:_
```Go
		// TODO(andrei): Figure out how to tolerate errors; we should put the events
		// back in the buffer (or not take them out of the buffer in the first
		// place) in hope that a new subscriber comes along.
```
___
*[`pkg/obs/event_exporter.go` line 427 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bERtCBAqVb3AEleBb:-N7bERtD40upKzu1RqDq:bqq8koi) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L427)):*
> ```Go
> ) *subscriber {
> 	sub := &subscriber{
> 		res:           s.resource,
> ```

nit:  should we do a sanity check here that the resource has been set?



_Code quote:_
```Go
		res:           s.resource,
```
___
*[`pkg/obs/event_exporter.go` line 487 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bF3vV9UHXK4iWzyu7:-N7bF3vV9UHXK4iWzyu8:bgmhsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L487)):*
> ```Go
> 	s.reset(ctx, sub)
> 	// TODO(andrei): It might be a good idea to return errors in some cases
> 	// (another subscriber coming, or quiescence).
> ```

Agreed, the subscriber should be told that information. It seems like the error passed to `subscriber.close` and sent along to `subscriber.stopC` isn't used currently? Maybe `runFlusher` should return an `error` that we can return here.



_Code quote:_
```Go
	// (another subscriber coming, or quiescence).
```
___
*[`pkg/obs/event_exporter.go` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bIkif1ZKdpOVQKwel:-N7bIkif1ZKdpOVQKwem:bqmybfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L501)):*
> ```Go
> 	ctx context.Context, bufs *eventsBuffers, flushInterval time.Duration,
> ) {
> 	defer close(sub.flusherDoneC)
> ```

Should we defer the closing of `sub.flushC` and `sub.doneC` too?



_Code quote:_
```Go
	defer close(sub.flusherDoneC)
```
___
*[`pkg/server/server.go` line 776 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7avZ3Q8Ir-46cnJgJB:-N7avZ3R7GJlIIGpZMNq:bbxxzln) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/server/server.go#L776)):*
> ```Go
> 	registry.AddMetricStruct(kvProber.Metrics())
> 
> 	obsServer := obs.NewEventServer(
> ```

nit: can you note the necessary call `SetResourceInfo` (and maybe `RegisterObsServer` as well) in a comment here?
___
*[`pkg/server/server.go` line 829 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7av9vE9dOqp_DZSauq:-N7av9vE9dOqp_DZSaur:bbtuvfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/server/server.go#L829)):*
> ```Go
> 		monitorAndMetrics:        sqlMonitorAndMetrics,
> 		settingsStorage:          settingsWriter,
> 		eventsServer:             obsServer,
> ```

nit: thoughts on calling it `obsServer` as well in the SQL Server for consistency? 



_Code quote:_
```Go
eventsServer:             obsServer,
```
___
*[`pkg/sql/event_log.go` line 120 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b3VQu7pJNsvFfO_cF:-N7b3VQv4QY8fSK-a54q:bj1fq9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L120)):*
> ```Go
> //     |
> //     +--> system.eventlog if not disabled by setting
> //		 |                   └ also the Obs Service, if connected
> ```

nit: switch to spaces for better alignment 



_Code quote:_
```Go
//		 |
```
___
*[`pkg/sql/event_log.go` line 487 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b2Loy4DQ4oXn4aQyW:-N7b2Loy4DQ4oXn4aQyX:b-3js48c) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L487)):*
> ```Go
> 	ctx context.Context,
> 	ex *InternalExecutor,
> 	eventsExporter obs.EventsExporter,
> ```

nit: Unused - should we use this below instead of `ex.s.cfg.EventsExporter`?



_Code quote:_
```Go
eventsExporter obs.EventsExporter,
```
___
*[`pkg/sql/event_log.go` line 603 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b7qSqCKmBnEzojkVf:-N7b7qSqCKmBnEzojkVg:b-t3501w) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L603)):*
> ```Go
> 	})
> 
> 	// In the common case where we have just 1 event, we want to skeep
> ```

I'm curious about what originally led to this optimization. 

It seems like the old code is intentionally avoiding loops, which we've introduced above in this PR in the primary code path. 

cc @knz - I see this was added by you a while back in https://github.com/cockroachdb/cockroach/commit/9318af940f86079628584ef146bb8e0bd71d505f. Do you by any chance remember if there were any adverse performance effects that led you to add this optimization? Or was it just done because it was an easy optimization to make here? I just want to make sure we're not inadvertently introducing any weird side-effects with these changes.



_Code quote:_
```Go
	// In the common case where we have just 1 event, we want to skeep
	// the extra heap allocation and buffer operations of the loop
	// below. This is an optimization.
```

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @andreimatei, and @dhartunian)

___
*[`-- commits` line 13 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7alIttBeJBYt6_jUG6):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

:fire: 
</blockquote></details>

See #84983.
___
*[`pkg/sql/event_log.go` line 603 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b7qSqCKmBnEzojkVf:-N7pJEI5Do1JIC6aIOY3:bbthc64) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L603)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

I'm curious about what originally led to this optimization. 

It seems like the old code is intentionally avoiding loops, which we've introduced above in this PR in the primary code path. 

cc @knz - I see this was added by you a while back in https://github.com/cockroachdb/cockroach/commit/9318af940f86079628584ef146bb8e0bd71d505f. Do you by any chance remember if there were any adverse performance effects that led you to add this optimization? Or was it just done because it was an easy optimization to make here? I just want to make sure we're not inadvertently introducing any weird side-effects with these changes.
</blockquote></details>

Yes this optimization is important if we actually write to the system.eventlog table -- it dramatically decreases the number of KV roundtrips when there are multiple events bundled together (such as what happen during a DROP CASCADE, certain priv changes etc).

It is less important if we do not write to the system.eventlog table (see #84983).

**Review by andreimatei — COMMENTED** [review]
Not quite done with the tests, but it's now close to what I'd merge.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-:-N8VemrQCS9ZgnZEcNEA:b-u5kwwb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @andreimatei, and @dhartunian)

___
*[`pkg/obs/event_exporter.go` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7arYT696mhSwzQshAo:-N8U-JZc5mitnsm63_fw:b2nih0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L43)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Cool - I could see us building cockroachdb receiver & exporter components for the otel collector some day to integrate with this . 
</blockquote></details>

yup
___
*[`pkg/obs/event_exporter.go` line 310 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bMbzHEu20hVk8P3ir:-N8U0Ogl2vIkqaXqFT4w:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L310)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: should we check/handle `!ok` with this map access?
</blockquote></details>

done
___
*[`pkg/obs/event_exporter.go` line 411 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bLUJjDou0ARrVMdxC:-N8U3-XADdUDStxB3EA9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L411)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

in the meantime, maybe we should log something here.
</blockquote></details>

done
___
*[`pkg/obs/event_exporter.go` line 427 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bERtCBAqVb3AEleBb:-N8U3OUA80_RPX_-Xwuz:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L427)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit:  should we do a sanity check here that the resource has been set?
</blockquote></details>

done
___
*[`pkg/obs/event_exporter.go` line 487 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bF3vV9UHXK4iWzyu7:-N8UJcb34Eck3HtxTUMH:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L487)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Agreed, the subscriber should be told that information. It seems like the error passed to `subscriber.close` and sent along to `subscriber.stopC` isn't used currently? Maybe `runFlusher` should return an `error` that we can return here.
</blockquote></details>

done
___
*[`pkg/obs/event_exporter.go` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bIkif1ZKdpOVQKwel:-N8UKBxvD_Iwbqqg5_HM:bv725vj) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L501)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Should we defer the closing of `sub.flushC` and `sub.doneC` too?
</blockquote></details>

It would not be correct to close `flushC` because `SendEvent` might be sending on it concurrently. But note that `SendEvent's` write is not blocking.
For `doneC` I think you meant `flushAndStopC`. That channel is also not for `runFlusher` to close; `runFlusher` is read-only for the subscriber (that's also enforced at the type level, so we can't `close` it even if we wanted to. `flushAndStopC` is the stopper  channel.
___
*[`pkg/server/server.go` line 776 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7avZ3Q8Ir-46cnJgJB:-N8ULWdA-TyKt3R4nw95:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/server/server.go#L776)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: can you note the necessary call `SetResourceInfo` (and maybe `RegisterObsServer` as well) in a comment here?
</blockquote></details>

done
___
*[`pkg/server/server.go` line 829 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7av9vE9dOqp_DZSauq:-N8UL_nL9Vixpa9p_Ehe:b-v3k769) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/server/server.go#L829)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: thoughts on calling it `obsServer` as well in the SQL Server for consistency? 
</blockquote></details>

renamed to `eventServer` everywhere
___
*[`pkg/sql/event_log.go` line 120 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b3VQu7pJNsvFfO_cF:-N8UTmil3zQBNgE8cOeg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L120)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: switch to spaces for better alignment 
</blockquote></details>

done
___
*[`pkg/sql/event_log.go` line 487 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b2Loy4DQ4oXn4aQyW:-N8UUKGtDti1cIva4ZVl:b-5g9o7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L487)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: Unused - should we use this below instead of `ex.s.cfg.EventsExporter`?
</blockquote></details>

yes, done
___
*[`pkg/sql/event_log.go` line 603 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b7qSqCKmBnEzojkVf:-N8UVeVI16s294nN8M_X:bqawhdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L603)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes this optimization is important if we actually write to the system.eventlog table -- it dramatically decreases the number of KV roundtrips when there are multiple events bundled together (such as what happen during a DROP CASCADE, certain priv changes etc).

It is less important if we do not write to the system.eventlog table (see #84983).
</blockquote></details>

I've added some loops, but I don't see anything scary in them.

File: pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql [hunk 1]
```diff
@@ -1,5 +1,13 @@
 -- +goose Up
-CREATE TABLE events();
+CREATE TABLE cluster_events(
+  timestamp TIMESTAMP NOT NULL,
+  id BYTES NOT NULL DEFAULT uuid_v4(),
+  cluster_id BYTES NOT NULL,
+  node_id INT NOT NULL,
+  event_type STRING NOT NULL,
+  event JSONB,
+  CONSTRAINT "primary" PRIMARY KEY (timestamp, id) USING HASH WITH BUCKET_COUNT = 16
```
Comment (ajwerner) [inline pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql:9]: These days you don't need the `WITH BUCKET_COUNT = 16`, it's the default. Also, this syntax is deprecated in favor of `CONSTRAINT "primary" PRIMARY KEY (timestamp, id) USING HASH WITH (bucket_count = 16)` if you did want to state it explicitly.

File: pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql [hunk 2]
```diff
@@ -1,5 +1,13 @@
 -- +goose Up
-CREATE TABLE events();
+CREATE TABLE cluster_events(
+  timestamp TIMESTAMP NOT NULL,
+  id BYTES NOT NULL DEFAULT uuid_v4(),
+  cluster_id BYTES NOT NULL,
+  node_id INT NOT NULL,
```
Comment (ajwerner) [inline pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql:6]: Are you interested in calling this `instance_id`? Should this have a `tenant_id`?

File: pkg/obsservice/obslib/ingest/ingest.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			}
+		}
+	}
+	return nil
+}
+
+const maxEventsPerStatement = 100
+
+// persistEventlogEvents writes "eventlog" events to the database.
+func persistEventlogEvents(
```
Comment (ajwerner) [inline pkg/obsservice/obslib/ingest/ingest.go:127]: Any interest in making yourself an interface to write the data to persistence for later dependency injection. Feels a little crazy to be passing `pgxpool.Pool` here, but maybe that's just me. It seems like you ought to make yourself a data model and then have some interfaces to the database to interact with that model.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @ajwerner, @andreimatei, and @dhartunian)

___
*[`pkg/obsservice/obslib/ingest/ingest.go` line 127 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7NgRlfQgrFEmRkUpRG-r4-127:-N8YzAy228SxCdghYjmp:b-t6ry4) ([raw file](https://github.com/cockroachdb/cockroach/blob/034059a950f283668fdf2f460329a6ce7b5d28db/pkg/obsservice/obslib/ingest/ingest.go#L127)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Any interest in making yourself an interface to write the data to persistence for later dependency injection. Feels a little crazy to be passing `pgxpool.Pool` here, but maybe that's just me. It seems like you ought to make yourself a data model and then have some interfaces to the database to interact with that model. 
</blockquote></details>

But what's wrong with passing in a pool (as a handle to a database)? Even for unit tests this works fine - i.e. I don't see tests using any sort of mock.
I'm not very interested in introducing more abstraction in this PR. Did you have an ORM in mind? I imagine we will probably end up some ORM for some event types (probably more for the needs of API endpoints reading the events, rather than for ingesting them), but I'd wait for that need to come (partially because I'll have to educate myself on ORMs when that moment comes). And it might not come for all event types.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @andreimatei, and @dhartunian)

___
*[`pkg/obsservice/obslib/ingest/ingest.go` line 127 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7NgRlfQgrFEmRkUpRG-r4-127:-N8Z0BBI03SQfJ88ddwT:bfg86z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/034059a950f283668fdf2f460329a6ce7b5d28db/pkg/obsservice/obslib/ingest/ingest.go#L127)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

But what's wrong with passing in a pool (as a handle to a database)? Even for unit tests this works fine - i.e. I don't see tests using any sort of mock.
I'm not very interested in introducing more abstraction in this PR. Did you have an ORM in mind? I imagine we will probably end up some ORM for some event types (probably more for the needs of API endpoints reading the events, rather than for ingesting them), but I'd wait for that need to come (partially because I'll have to educate myself on ORMs when that moment comes). And it might not come for all event types.
</blockquote></details>

Fair enough. Wait-and-see seems fine.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @ajwerner, @andreimatei, and @dhartunian)

___
*[`pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7NgRlfQgrFEmRkUpR1-r4-6:-N8ZHvbfBSUfRltvoazr:bhh24ju) ([raw file](https://github.com/cockroachdb/cockroach/blob/034059a950f283668fdf2f460329a6ce7b5d28db/pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql#L6)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Are you interested in calling this `instance_id`? Should this have a `tenant_id`? 
</blockquote></details>

renamed to `instance_id`
For tenants, the plan is for `cluster_id` to represent the tenant id, and have a separate table mapping tenant cluster to host clusters. This other table will be populated by a yet-to-be-written component that detects tenants coming (and also going?) into a host cluster that's being monitored, as well as SQL pods coming and going. FWIW, the component that watches for nodes coming and going in a host cluster is also yet-to-be-written.

FWIW, I see that in CRDB we don't talk about "tenant_id"; we just talk about cluster IDs. E.g. in `tenant.go`, we do `args.rpcContext.LogicalClusterID.Get()`.

Does this all sound right?
___
*[`pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql` line 9 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7NgRlfQgrFEmRkUpR1-r4-9:-N8ZHZc02yiOn6of5ljU:b-mm8nkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/034059a950f283668fdf2f460329a6ce7b5d28db/pkg/obsservice/obslib/migrations/sqlmigrations/0001_init.sql#L9)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

These days you don't need the `WITH BUCKET_COUNT = 16`, it's the default. Also, this syntax is deprecated in favor of `CONSTRAINT "primary" PRIMARY KEY (timestamp, id) USING HASH WITH (bucket_count = 16)` if you did want to state it explicitly. 
</blockquote></details>

I do prefer it explicit. I've switched to the parens syntax.

**Review by abarganier — APPROVED** [review]
Sorry for the review delay! :lgtm_strong: : 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-:-N9hgU1RAB4KcIgFZFrk:b2jd470)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @ajwerner, @andreimatei, @dhartunian, and @knz)

___
*[`pkg/obs/event_exporter.go` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7bIkif1ZKdpOVQKwel:-N9hfxUj1xSSdMAMXjpI:b-jfclqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/obs/event_exporter.go#L501)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It would not be correct to close `flushC` because `SendEvent` might be sending on it concurrently. But note that `SendEvent's` write is not blocking.
For `doneC` I think you meant `flushAndStopC`. That channel is also not for `runFlusher` to close; `runFlusher` is read-only for the subscriber (that's also enforced at the type level, so we can't `close` it even if we wanted to. `flushAndStopC` is the stopper  channel.
</blockquote></details>

I see, my mistake - thanks for explaining!
___
*[`pkg/sql/event_log.go` line 603 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-N7b7qSqCKmBnEzojkVf:-N9hgR-eA1pBnD-AhCp3:bcjx251) ([raw file](https://github.com/cockroachdb/cockroach/blob/71706fc83348c8cb027bd250c982d6f697fb7b00/pkg/sql/event_log.go#L603)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I've added some loops, but I don't see anything scary in them.
</blockquote></details>

Thanks for clarifying @knz  - lgtm

**Review by andreimatei — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84683#-:-N9nDp5eDDQmz3Q_0tHp:b-tlx7ko)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @ajwerner, @andreimatei, @dhartunian, and @knz)

**Pull request merged**